src/btree/bt_read.c

```diff
@@ -140,6 +141,242 @@ __instantiate_birthmarks(WT_SESSION_IMPL *session, WT_REF *ref)
+
+            if (lastupd != NULL) {
+                WT_ERR(__col_instantiate(session, current_recno, ref, &cbt, lastupd));
+
+                /* Remove the prepared record from LAS */
+                if (!birthmark_record && lastupd->prepare_state == WT_PREPARE_INPROGRESS) {
+                    cursor->prev(cursor);
+                    cursor->remove(cursor);
+                    cursor->next(cursor);
+                }
+
+                if (birthmark_record) {
+                    birthmarkp->instantiated = true;
+                    birthmark_record = false;
+                }
+                lastupd = NULL;
+            }
+            current_recno = recno;
+            break;
+        case WT_PAGE_ROW_LEAF:
+            if (current_key->size == las_key.size &&
+              memcmp(current_key->data, las_key.data, las_key.size) == 0) {
+                /*
+                 * For birthmark records insert only the birthmark record, and for other records,
+                 * insert the latest record for that key.
+                 */
+                if (birthmark_record)
+                    continue;
+                else
+                    break;
+            }
+
+            if (lastupd != NULL) {
+                WT_ERR(__row_instantiate(session, current_key, ref, &cbt, lastupd));
+
+                /* Remove the prepared record from LAS */
+                if (!birthmark_record && lastupd->prepare_state == WT_PREPARE_INPROGRESS) {
+                    cursor->prev(cursor);
+                    cursor->remove(cursor);
+                    cursor->next(cursor);
+                }
+
+                if (birthmark_record) {
+                    birthmarkp->instantiated = true;
+                    birthmark_record = false;
+                }
+                lastupd = NULL;
+            }
+            WT_ERR(__wt_buf_set(session, current_key, las_key.data, las_key.size));
+            break;
+        default:
+            WT_ERR(__wt_illegal_value(session, page->type));
+        }
+
+        if (ref->page_las->birthmarks_cnt > 0) {
+            for (; birthmark_cnt < ref->page_las->birthmarks_cnt; birthmark_cnt++) {
+                birthmarkp = ref->page_las->birthmarks + birthmark_cnt;
+                if (birthmarkp->instantiated)
+                    continue;
+
+                WT_ERR(__wt_compare(session, NULL, &las_key, &birthmarkp->key, &cmp));
+                if (cmp == 0) {
+                    WT_ERR(__create_birthmark_upd(session, birthmarkp, &upd));
+                    birthmark_record = true;
+                    break;
+                } else if (cmp < 0)
+                    /* LAS key is less than birthmark key, so no birthmark record for that key */
+                    break;
+                else
+                    /* LAS key is greater than birthmark key, check next set of birthmark records */
+                    continue;
+            }
+        }
+
+        if (upd == NULL) {
```

Comment:
Can you confirm what I think is happening here: We are allocating an update for each record in the lookaside. Then we discard that update when we move to the next update in las. When we reach the last record for that key, we initialize that update in memory.

If my understanding is correct, we can do two improvements here:
1. We can avoid allocating and freeing the updates we are not going to initialize in the memory.
2. We do not need to process updates for a key that are older than unstable_timestamp.

I think we can achieve both these cases by moving to the last record for a key in the lookaside, by searching near the UINT64_MAX for a key, that search should put the las cursor either on the newest update for that key, or one past on the next key.

Let me know if my understanding is wrong here.

Reply:
Used the search_near approach to improve the performance.